File: backends/go/internal/service/task_service.go
package service

import (
	"errors"
	"multi-lang-backend-go/internal/domain"
	"multi-lang-backend-go/internal/repository"
	"time"
)

var ErrTaskNotFound = errors.New("task not found")

type TaskService struct {
	repo *repository.Factory
}

func NewTaskService(repo *repository.Factory) *TaskService {
	return &TaskService{repo: repo}
}

func (s *TaskService) ListByUserID(userID string) ([]*domain.Task, error) {
	return s.repo.TaskRepository().FindByUserID(userID)
}

func (s *TaskService) Create(userID, title, description string) (*domain.Task, error) {
	task := &domain.Task{
		Title:       title,
		Description: description,
		UserID:      userID,
		CreatedAt:   time.Now(),
	}
	if err := s.repo.TaskRepository().Save(task); err != nil {
		return nil, err
	}
	return task, nil
}

func (s *TaskService) GetByID(id, userID string) (*domain.Task, error) {
	task, err := s.repo.TaskRepository().FindByID(id)
	if err != nil || task == nil {
		return nil, ErrTaskNotFound
	}
	if task.UserID != userID {
		return nil, ErrTaskNotFound
	}
	return task, nil
}

func (s *TaskService) Update(id, userID, title, description string) (*domain.Task, error) {
	existing, err := s.GetByID(id, userID)
	if err != nil {
		return nil, err
	}
	if title != "" {
		existing.Title = title
	}
	if description != "" {
		existing.Description = description
	}
	if err := s.repo.TaskRepository().Save(existing); err != nil {
		return nil, err
	}
	return existing, nil
}

func (s *TaskService) Delete(id, userID string) error {
	if _, err := s.GetByID(id, userID); err != nil {
		return err
	}
	return s.repo.TaskRepository().DeleteByID(id)
}

File: backends/java/src/main/java/com/example/multilang/service/TaskService.java
package com.example.multilang.service;

import com.example.multilang.config.RepositoryFactory;
import com.example.multilang.domain.Task;
import com.example.multilang.repository.TaskRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;

@Service
@RequiredArgsConstructor
public class TaskService {
    private final RepositoryFactory repositoryFactory;

    public List<Task> findByUserId(String userId) {
        TaskRepository repo = repositoryFactory.getTaskRepository();
        return repo.findByUserId(userId);
    }

    public Task create(String userId, String title, String description) {
        TaskRepository repo = repositoryFactory.getTaskRepository();
        Task task = Task.builder()
                .title(title)
                .description(description != null ? description : "")
                .userId(userId)
                .createdAt(Instant.now())
                .build();
        return repo.save(task);
    }

    public Task getById(String id, String userId) {
        TaskRepository repo = repositoryFactory.getTaskRepository();
        return repo.findById(id)
                .filter(t -> userId.equals(t.getUserId()))
                .orElseThrow(() -> new IllegalArgumentException("Task not found"));
    }

    public Task update(String id, String userId, String title, String description) {
        TaskRepository repo = repositoryFactory.getTaskRepository();
        Task existing = getById(id, userId);
        Task updated = Task.builder()
                .id(existing.getId())
                .title(title != null ? title : existing.getTitle())
                .description(description != null ? description : existing.getDescription())
                .userId(existing.getUserId())
                .createdAt(existing.getCreatedAt())
                .build();
        return repo.save(updated);
    }

    public void delete(String id, String userId) {
        getById(id, userId);
        repositoryFactory.getTaskRepository().deleteById(id);
    }
}
